waiter/src/waiter/scheduler/cook.clj

```diff
@@ -0,0 +1,540 @@
+                                           {:health-check-url health-check-url})
+                                false))))]
+      ;; Do not track unhealthy instances in the cache
+      (when-not health-result
+        (utils/atom-cache-evict healthy-instance-cache task-id))
+      health-result)))
+
+(defn- job->port
+  "Extracts the backend port from the job."
+  [job]
+  (if-let [backend-port (-> job :labels :backend-port)]
+    (Integer/parseInt backend-port)
+    (-> job :instances first :ports first)))
+
+(defn job-healthy?
+  "Returns true if the specified job has a healthy running instance."
+  [job]
+  (and (= "running" (-> job :status))
+       (= "running" (-> job :instances first :status))
+       (let [job-instance (-> job :instances first)
+             host (-> job-instance :hostname)
+             port (job->port job)
+             protocol (-> job :labels :backend-proto)
+             health-check-url (-> job :labels :health-check-url)
+             task-id (-> job-instance :task_id)]
+         (when (and host port protocol health-check-url)
+           (instance-healthy? task-id (str protocol "://" host ":" port health-check-url))))))
+
+;; Helper Methods
+
+(let [instance-expiry-adjustment-mins 5]
+  (defn create-job-description
+    "Create the Cook job description for a service."
+    [service-id service-description service-id->password-fn home-path-prefix instance-priority]
+    (let [{:strs [backend-proto cmd cpus health-check-url instance-expiry-mins mem metadata name ports
+                  run-as-user version]} service-description
+          job-uuid (str (UUID/randomUUID)) ;; TODO Use "less random" UUIDs for better Cook cache performance.
+          home-path (str home-path-prefix run-as-user)
+          environment (scheduler/environment service-id service-description service-id->password-fn home-path)
+          cook-backend-port (get metadata "cook-backend-port")
+          container-type (get metadata "container-type")
+          [image-namespace image-name-and-label] (str/split version #"/" 2)
+          [image-name image-label] (str/split (str image-name-and-label) #":" 2)
+          container-support-enabled? (and (seq container-type)
+                                          (seq image-namespace)
+                                          (seq image-name)
+                                          (seq image-label))]
+      (when cook-backend-port
+        (try
+          (when-not (pos? (Integer/parseInt cook-backend-port))
+            (throw (ex-info "cook-backend-port metadata parsed to a non-positive integer"
+                            {:cook-backend-port cook-backend-port})))
+          (catch ExceptionInfo ex
+            (throw ex))
+          (catch Exception ex
+            (throw (ex-info "cook-backend-port metadata must parse to a positive integer"
+                            {:cook-backend-port cook-backend-port} ex)))))
+      (when (seq container-type)
```

Comment:
Is `container-type` expected to be a sequence?

Reply:
Avoided using `seq` to check blank strings.